[Bugfix] [issue-21565] Fix the incompatibility issue with stream and named function calling when Thinking is disabled

hsliuustc0106 opened the pull request:
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

This PR fix bugs from #21565

## Test Plan

Online serving using the following command:
```
vllm serve Qwen/Qwen3-4B  --reasoning-parser qwen3 --enable-auto-tool-choice --tool-call-parser Hermes
```

Run the following code:
```Python
from openai import OpenAI


# Now, simulate a tool call
def get_current_weather(city: str, state: str, unit: "str"):
    return (
        "The weather in Dallas, Texas is 85 degrees fahrenheit. It is "
        "partly cloudly, with highs in the 90's."
    )


available_tools = {"get_current_weather": get_current_weather}

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

properties = {
    "city": {
        "type": "string",
        "description": "The city to find the weather for, e.g. 'San Francisco'",
    },
    "state": {
        "type": "string",
        "description": "the two-letter abbreviation for the state that the city is"
        " in, e.g. 'CA' which would mean 'California'",
    },
    "unit": {
        "type": "string",
        "description": "The unit to fetch the temperature in",
        "enum": ["celsius", "fahrenheit"],
    },
}

tools = [
    {
        "type": "function",
        "function": {
            "name": "get_current_weather",
            "description": "Get the current weather in a given location",
            "parameters": {
                "type": "object",
                "properties": properties,
                "required": ["city", "state", "unit"],
            },
        },
    }
]
messages = [
    {"role": "user", "content": "Hi! How are you doing today?"},
    {"role": "assistant", "content": "I'm doing well! How can I help you?"},
    {
        "role": "user",
        "content": (
            "Can you tell me what the temperate will be in Dallas, in fahrenheit?"
        ),
    },
]


def extract_reasoning_and_calls(chunks: list):
    reasoning_content = ""
    tool_call_idx = -1
    arguments = []
    function_names = []
    for chunk in chunks:
        if chunk.choices[0].delta.tool_calls:
            tool_call = chunk.choices[0].delta.tool_calls[0]
            if tool_call.index != tool_call_idx:
                tool_call_idx = chunk.choices[0].delta.tool_calls[0].index
                arguments.append("")
                function_names.append("")

            if tool_call.function:
                if tool_call.function.name:
                    function_names[tool_call_idx] = tool_call.function.name

                if tool_call.function.arguments:
                    arguments[tool_call_idx] += tool_call.function.arguments
        else:
            if hasattr(chunk.choices[0].delta, "reasoning_content"):
                reasoning_content += chunk.choices[0].delta.reasoning_content
    return reasoning_content, arguments, function_names


def main():
    client = OpenAI(
        api_key=openai_api_key,
        base_url=openai_api_base,
    )

    models = client.models.list()
    model = models.data[0].id

    print("----------Stream Generate With Named Function Calling--------------")

    tool_calls_stream = client.chat.completions.create(
        messages=messages,
        model=model,
        tools=tools,
        tool_choice={"type": "function", "function": {"name": "get_current_weather"}},
        stream=True,
        extra_body={"chat_template_kwargs": {"enable_thinking": False}}
    )

    chunks = list(tool_calls_stream)

    reasoning_content, arguments, function_names = extract_reasoning_and_calls(chunks)
    print(f"reasoning_content: {reasoning_content}")
    print(f"function name: {function_names[0]}")
    print(f"function arguments: {arguments[0]}")
    print("\n\n")


if __name__ == "__main__":
    main()
```

## Test Result
<img width="669" height="180" alt="image" src="https://github.com/user-attachments/assets/3a210b2e-268e-48fa-88e5-59a92ec33f99" />

## (Optional) Documentation Update

BruceW-07 committed:
[Bugfix]: Fix the incompatibility issue with stream and named function calling when Thinking is disabled

Signed-off-by: wangzi <3220100013@zju.edu.cn>

hsliuustc0106 committed:
Fix the incompatibility issue 21565

[Bugfix] Fix the incompatibility issue with stream and named function calling when Thinking is disabled

hsliuustc0106 commented:
@DarkLight1337  please help to ask some committer to review please

DarkLight1337 commented:
cc @chaunceyjiang @aarnphm

chaunceyjiang reviewed (commented):
Indeed, this is a bug, and I can also reproduce it locally.

My PR #19135 handles the cases where `tool_choice="auto"` and `tool_choice="required"`, but it indeed missed the case where `tool_choice="Forced Function"`.

Could you add a unit test for this case, similar to what was done in #19135?

hsliuustc0106 commented:
> Indeed, this is a bug, and I can also reproduce it locally.
> 
> My PR #19135 handles the cases where `tool_choice="auto"` and `tool_choice="required"`, but it indeed missed the case where `tool_choice="Forced Function"`.
> 
> Could you add a unit test for this case, similar to what was done in #19135?

will update later

chaunceyjiang commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -641,7 +643,12 @@ async def chat_completion_stream_generator(
+                                    list(output.token_ids)) or \
+                                    (res.prompt_token_ids and
+                                        reasoning_parser.is_reasoning_end(
+                                            list(res.prompt_token_ids)
```
nit: add some comments.

BruceW-07 committed:
fix the logic of accessing reasoning_end_arr

Signed-off-by: wangzi <3220100013@zju.edu.cn>

BruceW-07 committed:
add a unit test for completion with named function calling

Signed-off-by: wangzi <3220100013@zju.edu.cn>

hsliuustc0106 committed:
fix reasoning_end_error & add unit test

[Bugfix] [issue-21565] Fix the incompatibility issue with stream and named function calling when Thinking is disabled

BruceW-07 committed:
add comments

Signed-off-by: wangzi <3220100013@zju.edu.cn>

hsliuustc0106 committed:
add some comments

[Bugfix] [issue-21565] Fix the incompatibility issue with stream and named function calling when Thinking is disabled

hsliuustc0106 commented:
> Indeed, this is a bug, and I can also reproduce it locally.
> 
> My PR #19135 handles the cases where `tool_choice="auto"` and `tool_choice="required"`, but it indeed missed the case where `tool_choice="Forced Function"`.
> 
> Could you add a unit test for this case, similar to what was done in #19135?

unit tests added

chaunceyjiang reviewed (approved):
Thanks~

DarkLight1337 reviewed (approved):
Stamp
